### PR TITLE
Improve the developer docker experience

### PR DESCRIPTION
**Pull request by kmcnellis**
### Description
Improvements to the locally running developer experience
- Uses named volumes to separate sbt's docker artifacts from ones created by IDEs
- Adjusts the bin/sbt  and bin/dev-bash scripts to use docker compose (easier to maintain volume mappings)

- Update bin/run-dev to allow the local server to continue running in the background
- Add bin/stop-dev and bin/remove-dev to clean up the local server
- Improves output from local running scripts

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: bin/vscode-setup [hunk 1]
```diff
@@ -8,12 +8,22 @@ source bin/lib.sh
...
+if [ -e universal-application-tool-0.0.1/pom.xml ]; then
+  cp universal-application-tool-0.0.1/pom.xml \
+    universal-application-tool-0.0.1/pom.bkp
+fi
+
 bin/sbt makePom
 
-cp universal-application-tool-0.0.1/target/scala-*/*.pom \
-    universal-application-tool-0.0.1/pom.xml
+docker cp civiform-shell:/usr/src/universal-application-tool-0.0.1/target universal-application-tool-0.0.1/temp
```
Comment (goldblatt) [inline bin/vscode-setup:18]: do we not want to mount this in the dev compose?

File: bin/run-dev [hunk 2]
```diff
@@ -21,15 +21,15 @@ emulators::set_localstack_emulator_vars
...
-      "--azure")
-        emulators::ensure_only_one_cloud_provider_flag azure
-        emulators::set_azurite_emulator_vars
-        ;;
-
-      "--aws")
-        emulators::ensure_only_one_cloud_provider_flag aws
-        # Already defaulted to AWS.
-        ;;
+    "--azure")
```
Comment (goldblatt) [inline bin/run-dev:24]: mmm i think this is just a spacing change maybe we can revert this

File: bin/stop-dev [hunk 3]
```diff
@@ -2,4 +2,4 @@
 
 # DOC: Stop the running dev environment. Alias of 'docker-compose down'.
 
-docker-compose down
+docker-compose down
```
Comment (goldblatt) [inline bin/stop-dev:5]: i think you can revert the changes here!

File: bin/run-test-ci [hunk 4]
```diff
@@ -9,8 +9,9 @@ docker::ensure_unit_test_env
 docker run --rm \
   --network test-support_default \
   -v "$(pwd)/universal-application-tool-0.0.1:/usr/src/universal-application-tool-0.0.1" \
-  -v "$(pwd)/sbt_cache/coursier:/root/.cache/coursier" \
-  -v ~/.sbt:/root/.sbt \
-  -v ~/.ivy:/root/.ivy2 \
+  -v "civiform_target:/usr/src/universal-application-tool-0.0.1/target" \
+  -v "civiform_coursier-data:/root/.cache/coursier" \
```
Comment (goldblatt) [inline bin/run-test-ci:13]: ooo why are these prefaced with civiform_ instead of just being the volumes you specified below?

File: bin/dev-bash [hunk 5]
```diff
@@ -6,13 +6,6 @@ source bin/lib.sh
...
-  -it \
-  --rm \
-  -v /var/run/docker.sock:/var/run/docker.sock \
-  -v "$(pwd)/universal-application-tool-0.0.1:/usr/src/universal-application-tool-0.0.1" \
-  -v "$(pwd)/sbt_cache/coursier:/root/.cache/coursier" \
-  -v ~/.sbt:/root/.sbt \
-  -v ~/.ivy:/root/.ivy2 \
-  --entrypoint /bin/bash \
-  civiform-dev
+docker::run_shell_container
```
Comment (goldblatt) [inline bin/dev-bash:9]: so this switches to using the docker-compose up instead of run, not sure what that means? also does bin/dev-bash get used extensively?

File: bin/lib/docker.sh [hunk 6]
```diff
@@ -1,5 +1,41 @@
...
+#######################################
+# Starts a bash shell in the civiform-shell container
+#######################################
+function docker::run_shell_container() {
+  # Allocate a TTY for better output even though not strictly needed.
+  docker-compose \
+    -f docker-compose.yml \
+    -f docker-compose.dev.yml \
+    -f docker-compose.shell.yml \
+    up civiform \
```
Comment (goldblatt) [inline bin/lib/docker.sh:12]: yeah i guess im confused why we use civiform arg here and odn't use it other places?

**Review by goldblatt — CHANGES_REQUESTED** [review]
mostly just questions since im a lil unfamiliar with these changes; maybe good to have bion give a look to as i think he has more context, but in general this seems super cool thanks for taking a look into this and woot for speed improvements and learning more about docker!

File: bin/remove-dev [hunk 7]
```diff
@@ -0,0 +1,9 @@
+#! /usr/bin/env bash
+
+# DOC: Connect to the app running locally using Docker. Useful for manual testing.
+
+source bin/lib.sh
+
+echo "Deleting local civiform container"
+
+docker rm -f civiform
```
Comment (shanemc-goog) [inline bin/remove-dev:9]: Missing new line at EOF

File: docker-compose.shell.yml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+# # Mount code for hot reloading and cache for faster compile in dev mode.
+version: '3.1'
+
+services:
+  civiform:
+    command: ""
```
Comment (shanemc-goog) [inline docker-compose.shell.yml:6]: Empty string feels like a "magic value" can you comment on what behavior it's causing?  Is this meant to unset the base definition?

**Review by shanemc-goog — COMMENTED** [review]
wowee!  thank you. I'm not familiar with all this enough to really provide a lot of useful feedback and would +1 Susan's comment of letting Bion look at it.

File: bin/lib/docker.sh [hunk 9]
```diff
@@ -8,7 +44,7 @@
 #######################################
 function docker::run_dev_sbt_command() {
   # Allocate a TTY for better output even though not strictly needed.
-  docker run -it --rm \
+  docker run -it \
```
[the remark this answers is not in the document]
Comment (bion) [reply bin/lib/docker.sh:47]: --rm removes the container created by the `run` command, if we don't include `--rm` here this command will leave a new container around each time it's run

File: bin/pull-image [hunk 10]
```diff
@@ -8,6 +8,10 @@ if [[ -z "${USE_LOCAL_CIVIFORM}" ]]; then
   echo "Making sure we're up to date with the latest dev... "\
     "set environment variable USE_LOCAL_CIVIFORM=1 to skip"
 
-  docker pull civiform/civiform-dev:latest
+  docker pull -q civiform/civiform-dev:latest
   docker tag civiform/civiform-dev:latest civiform-dev
+
+else
+  echo "Using local civiform docker images"
+
```
Comment (bion) [inline bin/pull-image:16]:
```suggestion
```

File: docker-compose.shell.yml (hunk 8 above)
Comment (bion) [reply docker-compose.shell.yml:6]: since this file is used to override other docker-compose config files my guess this is clearing any value that was set in one of the others?

File: bin/lib.sh [hunk 11]
```diff
@@ -4,14 +4,17 @@
 # in the base bin/ directory should begin by sourcing it.
 
 if [[ "${KEEP_ORIGINAL_PWD}" != "true" ]]; then
-  pushd $(git rev-parse --show-toplevel) > /dev/null
+  pushd $(git rev-parse --show-toplevel) >/dev/null
```
Comment (bion) [inline bin/lib.sh:7]: please revert

File: bin/pull-image [hunk 12]
```diff
@@ -8,6 +8,9 @@ if [[ -z "${USE_LOCAL_CIVIFORM}" ]]; then
   echo "Making sure we're up to date with the latest dev... "\
     "set environment variable USE_LOCAL_CIVIFORM=1 to skip"
 
-  docker pull civiform/civiform-dev:latest
+  docker pull -q civiform/civiform-dev:latest
   docker tag civiform/civiform-dev:latest civiform-dev
+
```
Comment (bion) [inline bin/pull-image:13]:
```suggestion
```

File: bin/remove-dev [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+set +e # errors are OK
+
+echo "Stopping & deletng local civiform emulators"
+docker-compose -f docker-compose.yml -f docker-compose.dev.yml --profile emulator down
+echo "Stopping & deletng local civiform shell"
+docker-compose -f docker-compose.yml -f docker-compose.dev.yml --profile shell down
+echo "Stopping & deletng local civiform container"
+docker-compose -f docker-compose.yml -f docker-compose.dev.yml down
+
+docker network rm "${COMPOSE_PROJECT_NAME}_default"  2>/dev/null
```
Comment (bion) [inline bin/remove-dev:16]:
```suggestion
docker network rm "${COMPOSE_PROJECT_NAME}_default" 2> /dev/null
```

File: bin/run-test-ci [hunk 14]
```diff
@@ -2,15 +2,16 @@
 
 # DOC: Run the unit tests in CI mode.
 
+export COMPOSE_PROJECT_NAME=test-support
```
[the remark this answers is not in the document]
Comment (bion) [reply bin/run-test-ci:5]: looks like this is overriding the default value in bin/lib.sh

**Review by bion — CHANGES_REQUESTED** [review]
overall looks good! just a few nits

File: bin/attach-dev [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+source bin/lib.sh
+
+echo "Attaching to civiform.  Run Ctrl-P, followed by Ctrl-Q, to detach"
+echo "Ctrl-C will quit sbt, re-attach to restart"
+
+# Start a warmup request loop so that when the subsequent server startup is ready, we trigger lazy
+# loading its edit-refresh resources so save developer time.
+for i in {1..60}; do
+  sleep 5
+  curl --output /dev/null --silent --fail http://localhost:9000/loginForm && break
```
Comment (bion) [inline bin/attach-dev:14]: question (blocking): i think this might stop the execution of the script if it fails since `set -e`, this may need to be
```suggestion
  if curl --output /dev/null --silent --fail http://localhost:9000/loginForm; then
    break
  fi
```

File: bin/dev-bash (hunk 5 above)
Comment (bion) [reply bin/dev-bash:9]: It gets used a decent amount for debugging docker and env stuff

File: bin/lib.sh (hunk 11 above)
Comment (bion) [reply bin/lib.sh:7]: !

TIL, thank you

File: bin/remove-dev [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+set +e # errors are OK
+
+echo "Stopping & deleting local civiform emulators"
+docker-compose -f docker-compose.yml -f docker-compose.dev.yml --profile emulator down
+echo "Stopping & deletng local civiform shell"
+docker-compose -f docker-compose.yml -f docker-compose.dev.yml --profile shell down
+echo "Stopping & deletng local civiform container"
+docker-compose -f docker-compose.yml -f docker-compose.dev.yml down
+
+docker network rm "${COMPOSE_PROJECT_NAME}_default"  2>/dev/null
```
Comment (bion) [inline bin/remove-dev:16]:
```suggestion
docker network rm "${COMPOSE_PROJECT_NAME}_default" 2>/dev/null
```

**Review by bion — APPROVED** [review]

File: bin/attach-dev (hunk 15 above)
Comment (kmcnellis) [reply bin/attach-dev:14]: It doesn't stop the script (since this runs in the background), but it might very well be stopping the loop.

Changed

**Comment by kmcnellis** [comment]
Added compose instructions in https://github.com/civiform/docs/pull/4

File: bin/run-dev [hunk 17]
```diff
@@ -38,14 +38,12 @@ function set_args() {
...
-  "set environment variable USE_LOCAL_CIVIFORM=1 to skip"
-bin/pull-image > /dev/null
+bin/pull-image
 
 docker::set_network_name_dev
 
-# Start emulator.
-docker-compose --profile "${cloud_provider}" \
+# Start containers w/ specified emulator.
+docker compose --profile "${cloud_provider}" \
```
Comment (gwendolyngoetz) [inline bin/run-dev:46]: The script fails for me because of the  removal of the - in docker-compose.

Comment (bion) [reply bin/run-dev:46]: yeah, it requires everyone be on a more recent version of docker https://docs.docker.com/compose/cli-command/#install-on-linux

Comment (gwendolyngoetz) [reply bin/run-dev:46]: Can confirm. Installing docker-compose v2 as a plugin works great!

File: bin/run-dev [hunk 18]
```diff
@@ -54,15 +52,12 @@ docker-compose --profile "${cloud_provider}" \
...
-# Start a warmup request loop so that when the subsequent server startup is ready, we trigger lazy
-# loading its edit-refresh resources so save developer time.
-for i in {1..60}; do
-  sleep 5
-  curl --output /dev/null --silent --fail http://localhost:9000/loginForm && break
-done &
-
-# start everything else
-docker-compose \
+# Attach to the civiform container
```
Comment (shanemc-goog) [inline bin/run-dev:55]: nit: add a trailing period.

**Review by shanemc-goog — APPROVED** [review]

**Review by goldblatt — APPROVED** [review]
